ce/api/streamflow/watershed_streams.py

```diff
@@ -0,0 +1,217 @@
+        return
+    except GeospatialTypeError as e:
+        print("##### GeospatialTypeError")
+        abort(400, description="Station must be a WKT POINT: {}".format(e.message))
+
+    with get_time_invariant_variable_dataset(
+        sesh, ensemble_name, "flow_direction"
+    ) as flow_direction_ds, get_time_invariant_variable_dataset(
+        sesh, ensemble_name, "elev"
+    ) as elevation_ds, get_time_invariant_variable_dataset(
+        sesh, ensemble_name, "elevmin"
+    ) as elevation_min_ds, get_time_invariant_variable_dataset(
+        sesh, ensemble_name, "elevmax"
+    ) as elevation_max_ds, get_time_invariant_variable_dataset(
+        sesh, ensemble_name, "area"
+    ) as area_ds:
+        try:
+            return worker(
+                station_lonlat,
+                flow_direction=VicDataGrid.from_nc_dataset(
+                    flow_direction_ds, "flow_direction"
+                ),
+            )
+        except GeoDataGrid2DIndexError:
+            abort(
+                404,
+                description="Station lon-lat coordinates are not within the area "
+                "for which we have data.",
+            )
+
+
+def worker(
+    station_lonlat, flow_direction,
+):
+    """Compute the watershed streamflow.
+
+    This function exists to make these computations more easily testable.
+    (Specifically, data *files* are not required, only the relevant contents
+    of those files passed as `VicDataGrid` objects. `VicDataGrid`s are much easier to
+    construct for tests.)
+
+    :param station_lonlat: (tuple) Location of drainage point, (lon, lat)
+    :param flow_direction: (VicDataGrid) Flow direction grid
+    :return: (dict) representation for JSON response object; see watershed() for details
+    """
+
+    # Compute lonlats of watershed whose mouth is at `station`
+    # TODO: Refactor to accept a VicDataGrid?
+    direction_matrix = VIC_direction_matrix(
+        flow_direction.lat_step, flow_direction.lon_step
+    )
+
+    with Timer() as watershed_time:
+        watershed_xys = build_watershed_streams(
+            flow_direction.lonlat_to_xy(station_lonlat),
+            flow_direction.values,
+            direction_matrix,
+            debug=True,
+        )
+
+    # `watershed_lonlats` must be an ordered collection (not sets) because
+    # a multi line string is an array (python list) of linestrings
+    watershed_lonlats = [[]]
+    counter = 0
+
+    for frozen in watershed_xys:
+        for xy in frozen:
+            watershed_lonlats[counter].append(flow_direction.xy_to_lonlat(xy))
+        counter += 1
+        watershed_lonlats.append([])
+
+    if watershed_lonlats[-1] == []:
+        watershed_lonlats.remove([])
+
+    lines = MultiLineString(watershed_lonlats)
+
+    return {
+        "streams": geojson_feature(lines,),
+        "debug/test": {"watershed_streams": {"time": watershed_time.elapsed,}},
+    }
+
+
+def build_watershed_streams(target, routing, direction_map, debug=False):
```

Comment:
As we discussed on Zulip, this function is very similar to the watershed code. Can you abstract out the repeated/common part and share it between them? That would make any fixes apply to both, and reduce or eliminate duplication of testing code.

Reply:
If there are other commonalities, say in the worker or other higher level code, it _may_ be worth factoring that out too.